### PR TITLE
Add TOTP & HOTP Setup

### PR DESCRIPTION
**Pull request by therealssj**

### CODE REVIEW

File: src/Plugin/TfaSetup/TfaHotpSetup.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+ *   description = @Translation("TFA Hotp Setup Plugin"),
+ *   help_links = {
+ *    "Google Authenticator (Android/iPhone/BlackBerry)" = "https://support.google.com/accounts/answer/1066447?hl=en",
+ *    "Authy (Android/iPhone)" = "https://www.authy.com/thefuture#install-now",
+ *    "FreeOTP (Android)" = "https://play.google.com/store/apps/details?id=org.fedorahosted.freeotp",
+ *    "GAuth Authenticator (desktop)" = "https://github.com/gbraad/html5-google-authenticator"
+ *   }
+ * )
+ */
+class TfaHotpSetup extends TfaHotp implements TfaSetupInterface {
```
Comment (nerdstein) [inline src/Plugin/TfaSetup/TfaHotpSetup.php]: Is the logic of this class ONLY what GA_Login needs?

Why wouldn't this be called GALoginHotpSetup?

File: src/Plugin/TfaSetup/TfaTotpSetup.php [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ *   description = @Translation("TFA Totp Setup Plugin"),
+ *   help_links = {
+ *    "Google Authenticator (Android/iPhone/BlackBerry)" = "https://support.google.com/accounts/answer/1066447?hl=en",
+ *    "Authy (Android/iPhone)" = "https://www.authy.com/thefuture#install-now",
+ *    "FreeOTP (Android)" = "https://play.google.com/store/apps/details?id=org.fedorahosted.freeotp",
+ *    "GAuth Authenticator (desktop)" = "https://github.com/gbraad/html5-google-authenticator"
+ *   }
+ * )
+ */
+class TfaTotpSetup extends TfaTotp implements TfaSetupInterface {
```
Comment (nerdstein) [inline src/Plugin/TfaSetup/TfaTotpSetup.php]: See comment above about GA Login specific code and naming convention

Comment (nerdstein) [inline src/Plugin/TfaSetup/TfaTotpSetup.php]: Can any of this code be furnished by TFA and overridden by GA Login classes?

File: src/Plugin/TfaSetup/TfaHotpSetup.php (hunk 1 above)
Comment (therealssj) [inline src/Plugin/TfaSetup/TfaHotpSetup.php]: Aah true it should be GALoginHotpSetup

File: src/Plugin/TfaSetup/TfaTotpSetup.php (hunk 2 above)
Comment (therealssj) [inline src/Plugin/TfaSetup/TfaTotpSetup.php]: Do you mean like some sort of decorator?

Comment (nerdstein) [inline src/Plugin/TfaSetup/TfaTotpSetup.php]: I'm not sure, but if TFA can or should provide a base class with some/all of this code, the GA classes TfaTotpSetup and TfaHotpSetup would extend that class and provide ONLY it's overrides

Comment (therealssj) [inline src/Plugin/TfaSetup/TfaTotpSetup.php]: I have planned it for this week to explore the options. I was reading up on design patterns and had some ideas related to integrate decorators in other class but havent given this one much thought.
But given that most of the code is similar for the two cases, main difference being in `validate` function we can most probably provided a base class and extend upon it.

**Pull request merged**